### ISSUE TITLE
[pgbouncer doc] Suggested Pod Spread Constraint conflicts with default constraints 

### ISSUE THREAD

**Issue by Jean-Daniel** [CrunchyData/postgres-operator#3608]
In the pgbouncer configuration documentation (tutorial/connection-pooling/), there is a `Pod Spread Constraints` section explaining how to schedule the pods on different nodes.

But this section does not mention that PGO already set this constraints by default, as so, trying to set it on `pgbouncer` will make PGO trying to generate invalid `Deployment` configuration, and fails to deploy `pgbouncer`.

>  time="2023-03-16T10:42:27Z" level=error msg="Reconciler error" controller=postgrescluster controllerGroup=postgres-operator.crunchydata.com controllerKind=PostgresCluster error="failed to create typed patch object (mycluster/mycluster-pgbouncer; apps/v1, Kind=Deployment): .spec.template.spec.topologySpreadConstraints: duplicate entries for key [topologyKey=\"kubernetes.io/hostname\",whenUnsatisfiable=\"ScheduleAnyway\"]" file="internal/controller/postgrescluster/pgbouncer.go:518" func="postgrescluster.(*Reconciler).reconcilePGBouncerDeployment" name=mycluster namespace=myapp postgresCluster=mycluster/mycluster reconcileID=b83999e0-de50-4933-abe6-c9ead5e568f8 version=5.3.1-0

This section should be updated to explicitly mention that the defined constraints are append to the default constraints, and do not replace them.

**Comment by benjaminjb** [comment]
Hello, sorry you're running into this. I'm a little confused as to why this is happening -- I can't replicate it. Could you share the yaml for your postgrescluster? (Are you creating this cluster via Helm or via another way?)

**Comment by Jean-Daniel** [comment]
I'm using Kubernetes v1.24.9, but nonetheless, this documentation section should really mention the fact that by default, topology spread constraints are already generated to avoid deploying pods on the same node, and it is useless to replicate it.


```yaml
apiVersion: postgres-operator.crunchydata.com/v1beta1
kind: PostgresCluster
metadata:
  labels:
    app.kubernetes.io/instance: analytics
    app.kubernetes.io/name: analytics
    app.kubernetes.io/part-of: test
  name: analytics
  namespace: test
spec:
  backups:
    pgbackrest:
      global:
        repo1-retention-archive: '1'
        repo1-retention-archive-type: diff
        repo1-retention-diff: '7'
        repo1-retention-full: '2'
        repo1-retention-full-type: count
        repo1-retention-history: '365'
      repos:
      - name: repo1
        schedules:
          full: 30 1 * * 0
        volume:
          volumeClaimSpec:
            accessModes:
            - ReadWriteOnce
            resources:
              requests:
                storage: 2Gi
            storageClassName: rbd-retain
            volumeMode: Filesystem
  instances:
  - dataVolumeClaimSpec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
      storageClassName: rbd-retain
    name: database-1
    replicas: 1
    resources:
      limits:
        cpu: 4
        memory: 8Gi
      requests:
        cpu: 1
        memory: 4Gi
  - dataVolumeClaimSpec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
      storageClassName: rbd-retain
    name: database-2
    replicas: 1
    resources:
      limits:
        cpu: 4
        memory: 8Gi
      requests:
        cpu: 1
        memory: 4Gi
  monitoring:
    pgmonitor:
      exporter:
        configuration:
        - configMap:
            name: analytics-exporter-queries
        image: registry.developers.crunchydata.com/crunchydata/crunchy-postgres-exporter:ubi8-5.3.1-0
  openshift: false
  patroni:
    dynamicConfiguration:
      postgresql:
        parameters:
          synchronous_commit: 'off'
  postgresVersion: 14
  proxy:
    pgBouncer:
      affinity:
        podAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  postgres-operator.crunchydata.com/cluster: analytics
                  postgres-operator.crunchydata.com/data: postgres
              topologyKey: kubernetes.io/hostname
            weight: 100
      config:
        databases:
          '*': host=analytics-primary port=5432 dbname=analytics
          analytics-replicas: host=analytics-replicas port=5432 dbname=analytics max_db_connections=200
        global:
          default_pool_size: '200'
          max_client_conn: '16384'
          max_db_connections: '200'
          pool_mode: transaction
          server_tls_sslmode: require
      image: registry.developers.crunchydata.com/crunchydata/crunchy-pgbouncer:ubi8-1.18-0
      replicas: 1
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            postgres-operator.crunchydata.com/cluster: analytics
            postgres-operator.crunchydata.com/role: pgbouncer
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: ScheduleAnyway

```

**Comment by benjaminjb** [comment]
Ah, now I see it and you're right: we set a default for `topologySpreadConstraints` and if you try to add your own (to either the postgres, pgbouncer, or pgbackrest pods), we _append_ the constraints rather than merge them. 

(Off the top of my head right now, I wonder what would be the desired merge approach? Maybe it wouldn't even be a merge, but a total overwrite / preference for the custom settings? Right now I'm leaning towards that: if you need the defaults plus whatever customization, you can always add all of that yourself.)

As it stands, there is a way to prevent the defaults `topologySpreadConstraints ` from being added, which is to set the `spec.disableDefaultPodScheduling` to `true`. This is mentioned once in the docs [here](https://access.crunchydata.com/documentation/postgres-operator/5.3.1/architecture/scheduling/#customization), so I'll add a ticket to our backlog to add that documentation to the other places we talk about spread constraints. 

(One thing to be aware of there is that setting is global for the cluster, so it will affect pgbouncer AND postgres and pgbackrest pods, which should also be in the documentation.)

So to sum up:

(1) thanks for pointing out the documentation issue, I will add it to the backlog;
(2) if you want to add custom `topologySpreadConstraints` that conflict with the default, set `spec.disableDefaultPodScheduling` to `true`;
(3) if you set `spec.disableDefaultPodScheduling` to `true` but want the `topologySpreadConstraints` on all the pods, you'll have to add them as custom settings in the correct spots in the CR, like `spec.backups.pgbackrest.repoHost.topologySpreadConstraints`, `spec.instances[index].topologySpreadConstraints`, and `spec.proxy.pgBouncer.topologySpreadConstraints`.

**Comment by benjaminjb** [comment]
OK, I have captured this request in the backlog and hope the above comment helps. Thanks for pointing out this gap in the documentation!

I'm going to close this ticket now, but if you want to continue the conversation, please reopen it.
